backend/typescript/services/implementations/campService.ts

```diff
@@ -791,6 +801,117 @@ class CampService implements ICampService {
     }
   }
 
+  async updateCampSessionsByIds(
+    campId: string,
+    updatedCampSessions: Array<UpdateCampSessionsDTO>,
+  ): Promise<Array<CampSessionDTO>> {
+    let dbSession = null;
+    try {
+      const campSessionIds = updatedCampSessions.map((session) => session.id);
+      dbSession = await mongoose.startSession();
+      dbSession?.startTransaction();
+
+      const oldCampSessions = await MgCampSession.find({
+        _id: { $in: campSessionIds },
+        camp: { _id: campId },
+      });
+
+      if (oldCampSessions.length !== campSessionIds.length) {
+        throw new Error(
+          `Failed to update camp sessions. Could not find all camp sessions in ${campSessionIds} belonging to camp with id ${campId}`,
+        );
+      }
+
+      const camp = await MgCamp.findById(campId);
+      if (!camp) {
+        throw new Error(`Could not find camp with id=${campId}`);
+      }
+
+      const newCampSessions = await Promise.all(
+        updatedCampSessions.map(async (session) => {
+          if (camp.active) {
+            const oldCampSession: CampSession | null = await MgCampSession.findById(
+              session.id,
+            );
+            if (oldCampSession) {
+              if (
+                session.capacity &&
+                oldCampSession.campers &&
+                session.capacity < oldCampSession.campers.length
+              ) {
+                throw new Error(
+                  `Cannot decrease capacity to current number of registered campers. Requested capacity change: ${session.capacity}, current number of registed campers: ${oldCampSession.campers.length}`,
+                );
+              }
+              if (
+                session.dates &&
+                oldCampSession.dates.length !== session.dates.length
+              ) {
+                throw new Error(
+                  `Cannot change the number of dates for an active camp. Requested dates length: ${session.dates.length}, current number of dates ${oldCampSession.dates.length}`,
+                );
+              }
+            } else {
+              throw new Error(
+                `Failed to update camp sessions with ids=${campSessionIds}. CampSession with campSessionId ${session.id} and campId ${campId} not found.`,
+              );
+            }
+          }
+
+          const newCampSession: CampSession | null = await MgCampSession.findByIdAndUpdate(
+            session.id,
+            {
+              ...(session.capacity && { capacity: session.capacity }),
+              ...(session.dates && {
+                dates: session.dates.sort(
+                  (dateA, dateB) =>
+                    new Date(dateA).getTime() - new Date(dateB).getTime(),
+                ),
+              }),
+            },
+            { runValidators: true, new: true },
```

Comment:
I suspect it makes a new document if `findByIdAndUpdate` doesn't find a document with the given `session.id`. But this should never be the case no?

Reply:
`[options.new=false] «Boolean» if true, return the modified document rather than the original`

https://mongoosejs.com/docs/api.html#model_Model-findByIdAndUpdate

Reply:
thanks, looks good